tikv/deadlock.go

```diff
@@ -0,0 +1,312 @@
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	stream, err := deadlockPb.NewDeadlockClient(cc).Detect(ctx)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+	dtConn := &detectConn{
+		stream: stream,
+		ctx:    ctx,
+		cancel: cancel,
+	}
+	return dtConn, nil
+}
+
+func (c *detectConn) stop() {
+	c.cancel()
+}
+
+func (c *detectConn) send(req *deadlockPb.DeadlockRequest) error {
+	return c.stream.Send(req)
+}
+
+func (c *detectConn) recv() (*deadlockPb.DeadlockResponse, error) {
+	return c.stream.Recv()
+}
+
+func NewDetectClient() *detectClient {
+	return &detectClient{}
+}
+
+func (c *detectClient) getConn(addr string) (*detectConn, error) {
+	c.RLock()
+	if c.conn != nil {
+		return c.conn, nil
+	}
+	c.RUnlock()
+	newConn, err := NewDetectConn(addr)
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("detectClient connected to addr=%v", addr)
+	c.Lock()
+	defer c.Unlock()
+	if c.conn != nil {
+		newConn.stop()
+		return c.conn, nil
+	}
+	c.conn = newConn
+	return newConn, nil
+}
+
+func (c *detectClient) send(addr string, req *deadlockPb.DeadlockRequest) error {
+	conn, err := c.getConn(addr)
+	if err != nil {
+		return err
+	}
+	err = conn.send(req)
+	if err == nil {
+		return nil
+	}
+	conn.stop()
+	c.Lock()
+	if conn == c.conn {
+		log.Error("detectClient failed to send, invalidate current conn, err=%v", err)
+		c.conn = nil
+	}
+	c.Unlock()
+	return err
+}
+
+func (c *detectClient) recv(addr string) (*deadlockPb.DeadlockResponse, error) {
+	conn, err := c.getConn(addr)
+	if err != nil {
+		return nil, err
+	}
+	res, err := conn.recv()
+	if err == nil {
+		return res, nil
+	}
+	conn.stop()
+	c.Lock()
+	if conn == c.conn {
+		log.Error("detectClient failed to recv, invalidate current conn, err=%v", err)
+		c.conn = nil
+	}
+	c.Unlock()
+	return nil, err
+}
+
+func NewDeadlockDetector(waiterMgr *lockwaiter.Manager) *DeadlockDetector {
+	chSize := 1000
+	newDetector := &DeadlockDetector{
+		detector: deadlock.NewDetector(),
+		role:     Follower,
+		client:   NewDetectClient(),
+		sendCh:   make(chan *deadlockPb.DeadlockRequest, chSize),
+		waitMgr:  waiterMgr,
+	}
+	return newDetector
+}
+
+func (dt *DeadlockDetector) setLeaderStoreMeta(newMeta *metapb.Store) {
+	dt.mu.Lock()
+	defer dt.mu.Unlock()
+	dt.leaderStoreMeta = newMeta
+}
+
+func (dt *DeadlockDetector) getLeaderStoreMeta() *metapb.Store {
+	dt.mu.RLock()
+	defer dt.mu.RUnlock()
+	return dt.leaderStoreMeta
+}
+
+func (dt *DeadlockDetector) SendReqLoop() {
+	var (
+		leaderAddr string
+		err        error
+		refreshErr error
+	)
+	for {
+		req := <-dt.sendCh
+		leaderAddr = dt.leaderStoreMeta.GetAddress()
+		err = dt.client.send(leaderAddr, req)
+		if err != nil {
+			log.Warnf("send req to addr=%v failed, err=%v", leaderAddr, err)
+			refreshErr = dt.refreshFirstRegionLeader()
+			if refreshErr != nil {
+				log.Errorf("refresh leader addr failed, err=%v", refreshErr)
+			}
+			time.Sleep(3 * time.Second)
+			continue
```

Comment:
The failed `req` will be discarded when retry?

Reply:
Yes, the request will be timeout and the pessimistic lock error will be reported instead of possible deadlock error, tidb may retry the pessimistic lock request